azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/core/legacy/LegacyRefreshTokenCredentialFactory.java

```diff
@@ -0,0 +1,128 @@
+import com.azure.resourcemanager.AzureResourceManager;
+import com.azure.resourcemanager.resources.models.Tenant;
+import com.microsoft.aad.adal4j.AuthenticationContext;
+import com.microsoft.aad.adal4j.AuthenticationResult;
+import com.microsoft.aad.msal4j.IAuthenticationResult;
+import com.microsoft.azure.AzureEnvironment;
+import com.microsoft.azure.toolkit.lib.auth.exception.LoginFailureException;
+import com.microsoft.azure.toolkit.lib.common.exception.AzureToolkitRuntimeException;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.reflect.FieldUtils;
+import reactor.core.publisher.Mono;
+
+import java.net.MalformedURLException;
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.stream.Collectors;
+
+public class LegacyRefreshTokenCredentialFactory {
+    public static Mono<TokenCredential> createRefreshTokenCredential(AzureEnvironment environment, TokenCredential tokenCredential) {
+        return tokenCredential.getToken(new TokenRequestContext()
+                .addScopes(ScopeUtil.resourceToScopes(environment.managementEndpoint())))
+                .map(accessToken -> {
+                    IAuthenticationResult result = ((MsalToken) accessToken).getAuthenticationResult();
+                    String refreshTokenFromResult;
+                    try {
+                        refreshTokenFromResult = (String) FieldUtils.readField(result, "refreshToken", true);
+                    } catch (IllegalAccessException e) {
+                        throw new AzureToolkitRuntimeException("Cannot read refreshToken from IAuthenticationResult.");
+                    }
+                    if (StringUtils.isBlank(refreshTokenFromResult)) {
+                        throw new AzureToolkitRuntimeException("Fail to get refresh token.");
+                    }
+
+                    List<String> tenantIds = listTenantIds(environment, tokenCredential);
+                    if (CollectionUtils.isEmpty(tenantIds)) {
+                        throw new AzureToolkitRuntimeException("There are no tenants in your account.");
+                    }
+
+                    return LegacyRefreshTokenCredentialFactory.fromRefreshToken(
+                            environment, IdentityConstants.DEVELOPER_SINGLE_SIGN_ON_ID, tenantIds.get(0), refreshTokenFromResult);
+                });
+    }
+
+    public static TokenCredential fromRefreshToken(AzureEnvironment env, String clientId, String tenantId, String refreshToken) {
+        return buildRefreshTokenCredential(env, clientId, tenantId, refreshToken);
+    }
+
+    private static TokenCredential buildRefreshTokenCredential(AzureEnvironment env, String clientId, String tenantId, String refreshToken) {
```

Comment:
`buildRefreshTokenCredential` from another refreshToken?
rename to `refreshTokenCredential`?